Toy2/src/Calculator.java

```diff
@@ -0,0 +1,209 @@
+import math.*;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.*;
+
+public class Calculator {
+
+    private final List<String> stringInput; // 파싱된 입력값을 저장하는 리스트
+
+    /*
+     enum을 사용하게되면 Map<String, MathSign> operations = new HashMap<>();
+     이런 형태를 써야하나? 그러면 연산자 인터페이스를 구현한 클래스들은 어떻게 사용하지?..
+    */
+    private final Map<String, Calculation> operations = new HashMap<>(); // 연산자와 연산 클래스를 매핑하는 맵
+
+    public Calculator(String input) {
+        InputParser parser = new InputParser(input);
+        this.stringInput = parser.getParsedResult();
+        validateState();
+        initializeOperations();
+    }
+    // 연산자와 그에 대응하는 연산 클래스를 매핑
+    private void initializeOperations() {
+
+
+        //ENUM을 사용하면 궁극적인 형태는??이렇게?? -> operations.put(MathSign.PLUS.getValue(), MathSign.PLUS);
+        operations.put(MathSign.PLUS.getValue(), new Plus());
+        operations.put(MathSign.MINUS.getValue(), new Minus());
+        operations.put(MathSign.MULTIPLY.getValue(), new Multiple());
+        operations.put(MathSign.DIVIDE.getValue(), new Divide());
+        operations.put(MathSign.REMAINDER.getValue(), new Remainder()); // 나머지 연산 추가
+        // 필요한 경우 다른 연산자들도 이곳에 추가 가능
+    }
+
+    //연산 가능여부의 검증
+    private void validateState() {
+        long parenthesisBalance = 0; // 괄호의 균형을 검증하기 위한 변수
+        boolean lastWasOperator = true; // 마지막 토큰이 연산자인지 검증하기 위한 변수
+
+        //for문을 이용한부분을 메소드로 뺴고싶은데 지역변수가 존재하여 여기가 최선일것같다 다른방법이있을까?
+        for (int i = 0; i < stringInput.size(); i++) {
+
+            String token = stringInput.get(i);
+
+            if (token.equals("(")) {
```

Comment:
간단한 기법이지만, equals를 사용할 때에는 절대로 null이 되지 않는 리터럴에서 equals메서드를 시작해주는게 좋습니다. 의외로 이런 부분때문에 NPE가 자주 발생하니 아래와 같이 사용하는 습관을 들여주세요.

```suggestion
            if ("(".equals(token)) {
```